servant-client-ghcjs: Support binary requests

FPtje opened the pull request:
Introduces support for both sending and receiving binary data in `servant-client-ghcjs`.

Previously, the conversion to String before going to Bytestring would fail with any requests that sent or received binary data. With ArrayBuffers in the backend, sending Binary encoded data now works.

Also added a test case to https://github.com/FPtje/servant-client-ghcjs-test to test both sending and receiving of Binary data. It works and the tests don't show any regression.

One important note: I see that some recent version of servant has added a `streamingRequest` to the `RunClient`class. It is't implemented in servant-client-ghcjs, and this PR doesn't add it.

alpmestan commented:
@arianvp Thanks for offering to look into it. :-)

phadej commented:
`fail`, not `error`. Even better would be to `throwIO` with custom exception type.

phadej commented on servant-client-ghcjs/src/Servant/Client/Internal/XhrClient.hs:
```diff
@@ -217,42 +224,47 @@ setHeaders xhr request = do
+      RequestBodyLBS x -> return $ mBody $ BL.toStrict x
       RequestBodyBS x -> return $ mBody x
-      RequestBodyBuilder _ x -> return $ mBody $ toLazyByteString x
+      RequestBodyBuilder _ x -> return $ mBody $ BL.toStrict $ toLazyByteString x
```
As we on streaming case, we probably should add a note that `servant-client-ghcjs` doesn't respect any "constant space request body streaming" undertakings. (haddocks of `ClientM` and `client`?)

FPtje commented:
Updated: Rebased on latest master, added haddocks to `client` and `clientM` mentioning that streaming isn't supported. `streamingRequest` does a `throwIO` with a custom exception.

arianvp commented:
LGTM @phadej ?

arianvp commented:
Bump. Is this okay to merge?
